### PR TITLE
Avoid nested quantifiers with overlapping character space on git url parsing (#1902

### PR DESCRIPTION
**Pull request by finswimmer**
This PR started as a fix for #1902 to avoid nested quantifiers with overlapping character space during git url parsing.

It' result is a rework of the pattern matching and adding a test for `parse_url` to make this part more reliable.

Fixes: #1902

### CODE REVIEW

**Comment by k4nar** [comment]
Is there a coverage in unit test for those regexp? It looks like there should be :smile: .

**Comment by sdispater** [comment]
@k4nar Do you mean adding the cases mentioned in the original issue #1902? If so, I concur that we should add tests that cover these.

File: poetry/vcs/git.py [hunk 1]
```diff
@@ -22,7 +22,7 @@
     re.compile(
         r"^(git\+)?"
         r"(?P<protocol>https?|git|ssh|rsync|file)://"
-        r"(?:(?P<user>.+)@)*"
+        r"(?:(?P<user>\w+)@)*"
```
Comment (sdispater) [inline poetry/vcs/git.py]:
```suggestion
        r"(?:(?P<user>[a-zA-Z0-9_.-])@)*"
```

I think we only want standard characters here with the addition of `_`, `.`  and `-`. If you think `\w` is still needed, however I don't think so, adding those three characters is still needed.

Comment (finswimmer) [reply poetry/vcs/git.py]: What ever works is fine for me :) 

I thinks we should use the same pattern for the same groups. For example `resource` is one time `[a-z0-9_.-]*` and one time `[\w.\-]+`. But I'm not totally sure, what is allowed in each group. Any suggestions?

Comment (sdispater) [reply poetry/vcs/git.py]: I agree we should have consistency between the different regexps. I think for `resource`  `[a-z0-9_.-]+` should be enough.

Comment (finswimmer) [reply poetry/vcs/git.py]: I did some rework to be consistent within the pattern. I also added tests for `parse_url` which uncovered some problems with regexes. Now only the first of the two pattern for matching "protocol://" and for matching "user@" are used for our test data. Do you think we can remove the two other? The less regex we need, the better :smiley:

**Comment by sdispater** [comment]
@finswimmer Unless I'm mistaken, the affected strings mentioned (like `"https://" + "@" * 64 + "!"`) in #1902 are not present in the tests. I think we should add them to avoid regressions.

**Comment by finswimmer** [comment]
@sdispater I've added a test for this. But something went wrong with the checks on MacOS?

**Comment by sdispater** [comment]
@finswimmer Yes. I don't know what happened since nothing has changed, but it seems that the cached venv for MacOS is missing pip somehow. And since there is not easy way to clear the Github Actions cache at the moment I will need to make a PR to invalidate it.

**Comment by sdispater** [comment]
@finswimmer Cache issues should be fixed on master. You will have to rebase your branch to pick up the fixes

**Pull request merged**